### PR TITLE
Multi core

### PR DESCRIPTION
**Pull request by flopezag**
#### REVIEWERS
@jason-fox 

#### DESCRIPTION
- Update documentation about multi-threading functionality
- Correct indentation in installation guide content as well some links
- Translate into the Node Lib the execution of multi-threading and single-thread execution
- Change in the commonConfig to manage the new ENV variable IOTA_MULTI_CORE
- Added multiCore variable in config file
- Unit tests to check all the possibilities between ENV variable: IOTA_MULTI_CORE and config variable: multiCore
- Correction some typos

#### TESTING
Unit tests and Integrated with IoTAgent - JSON

### CODE REVIEW

File: lib/commonConfig.js [hunk 1]
```diff
@@ -305,6 +307,11 @@ function processEnvironmentVariables() {
         config.autocast = process.env.IOTA_AUTOCAST === 'true';
     }
 
+    if (process.env.IOTA_MULTI_CORE) {
+        config.multiCore = process.env.IOTA_MULTI_CORE === 'true';
+    } else {
```
Comment (jason-fox) [inline lib/commonConfig.js:312]: The else clause is unnecessary -  config.multiCore will default to falsy

File: lib/fiware-iotagent-lib.js [hunk 2]
```diff
@@ -81,8 +82,6 @@ function doActivate(newConfig, callback) {
         commandRegistry,
         securityService;
 
-    logger.format = logger.formatters.pipe;
```
Comment (jason-fox) [inline lib/fiware-iotagent-lib.js:84]: Why has logger been removed?

File: lib/fiware-iotagent-lib.js [hunk 3]
```diff
@@ -114,20 +113,6 @@ function doActivate(newConfig, callback) {
 
     config.setConfig(newConfig);
 
-    logger.getContext = function domainContext() {
```
Comment (jason-fox) [inline lib/fiware-iotagent-lib.js:117]: Why has logger been removed?

File: lib/fiware-iotagent-lib.js [hunk 4]
```diff
@@ -239,6 +224,76 @@ function deactivate(callback) {
     contextServer.stop(callback);
 }
 
+function executeIoTAgent(newConfig, iotAgent) {
```
Comment (jason-fox) [inline lib/fiware-iotagent-lib.js]: Add an additional parameter for the logging message

File: lib/fiware-iotagent-lib.js [hunk 5]
```diff
@@ -239,6 +224,76 @@ function deactivate(callback) {
...
+    iotAgent.start(newConfig, function (error) {
+        if (error) {
+            logger.error(context, 'Error starting JSON IoT Agent: [%s] Exiting process', error);
```
Comment (jason-fox) [inline lib/fiware-iotagent-lib.js]: This can't say JSON - it should be passed in not hard coded

File: lib/fiware-iotagent-lib.js [hunk 6]
```diff
@@ -239,6 +224,76 @@ function deactivate(callback) {
...
+        } else {
+            logger.info(context, 'JSON IoT Agent started');
```
Comment (jason-fox) [inline lib/fiware-iotagent-lib.js]: Same here

File: lib/fiware-iotagent-lib.js [hunk 7]
```diff
@@ -239,6 +224,76 @@ function deactivate(callback) {
...
+            op: domainObj.op,
+            srv: domainObj.service,
+            subsrv: domainObj.subservice,
+            msg: domainObj.msg,
+            comp: config.componentName || 'IoTAgent'
+        };
+    };
+
+    // If if was configured multi-core and the value is true
+    if (config.getConfig().multiCore === true) {
```
Comment (jason-fox) [inline lib/fiware-iotagent-lib.js:256]: I think `=== true` is redundant for a truthy test, but clearer. Is this style followed elsewhere?

File: test/unit/general/config-multi-core-test.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        });
+
+        it('should load the correct configuration parameter with value \'false\'', function(done) {
+            config.setConfig(iotAgentConfig);
+            config.getConfig().multiCore.should.equal(false);
+            done();
+        });
+    });
+
+    describe('When the IoT Agent is not started with Multi-Core environment variable and it is configured with ' +
```
Comment (jason-fox) [inline test/unit/general/config-multi-core-test.js:202]: Not sure if other ENV variables do a strict `true` check or merely truthy. This test would be redundant

File: doc/installationguide.md [hunk 9]
```diff
@@ -1,208 +1,286 @@
...
+Some of the configuration parameters can be overridden with environment variables, 
+to ease the use of those parameters with container-based technologies, like Docker, 
+Heroku, etc...
+
+The following table shows the accepted environment variables, as well as the 
+configuration parameter the variable overrides.
 
 | Environment variable      | Configuration attribute             |
 |:------------------------- |:----------------------------------- |
-| IOTA_CB_URL               | contextBroker.url                   |
```
Comment (jason-fox) [inline doc/installationguide.md:169]: Is this just sorting?

File: test/unit/general/config-multi-core-test.js [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+ * or (at your option) any later version.
+ *
+ * fiware-iotagent-lib is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with fiware-iotagent-lib.
+ * If not, seehttp://www.gnu.org/licenses/.
```
Comment (jason-fox) [inline test/unit/general/config-multi-core-test.js]: Missing space

File: doc/installationguide.md [hunk 11]
```diff
@@ -1,208 +1,286 @@
 # Installation & Administration Guide
 
 ## Configuration
-The `activate()` function that starts the IoT Agent receives as single parameter with the configuration for the IoT Agent.
-The Agent Console reads the same configuration from the `config.js` file.
+The `activate()` function that starts the IoT Agent receives as single parameter
```
Comment (jason-fox) [inline doc/installationguide.md:4]: Either all the indent stuff and tidying should be a separate PR or the real PR should explain that it it all whitespace apart from:

> * **multiCore**: When enabled, the IoT Agents runs in multi-thread environment to 
take advantage of multi-core systems. It allows two values `true` or `false`. This attribute 
is optional with default to false, which means that the IoTAgent runs in a single thread. 

And the ENV variable definition

File: lib/commonConfig.js (hunk 1 above)
Comment (flopezag) [reply lib/commonConfig.js:312]: I'm not sure. There are two options:
- Define the multi-core through environment variables, or
- Define it through the config.js file

The first choice is the environment variable, if we have not defined it we get if from the config file. 

Now the procedure is the following:
- If neither IOTA_MULTI_CORE or config.multiCore are defined the value assigned to config.multiCore is false.
- If IOTA_MULTI_CORE is defined (true or false), we take the value from the env. variable whatever value we have in the config.multiCore. Keep in mind that we check if it is true or false. If the value is false or whatever other value, we get the value false (line 311).
- if IOTA_MULTI_CORE is not defined, then we have the option to get the value from the config.js file. then we have the option, it is defined true so the value is true. It is defined false or whatever other value, then the selected value is false. It is not defined, the normal case then the value is also false.

File: lib/fiware-iotagent-lib.js (hunk 2 above)
Comment (flopezag) [reply lib/fiware-iotagent-lib.js:84]: Not really removed, moved to the startServer() (see line 241). The reason is that I need the configuration of the logger in a first iteration in order to write properly the logging message.

File: lib/fiware-iotagent-lib.js (hunk 3 above)
Comment (flopezag) [reply lib/fiware-iotagent-lib.js:117]: Idem to the previous comment, see startServer() function.

File: lib/fiware-iotagent-lib.js (hunk 4 above)
Comment (flopezag) [reply lib/fiware-iotagent-lib.js]: I do not understand what do you mean.

File: lib/fiware-iotagent-lib.js (hunk 5 above)
Comment (flopezag) [reply lib/fiware-iotagent-lib.js]: Shit, you are right this is the node lib. You mean that the function (error) just pass the error to the up functions (toward startServer) and from there to the corresponding IoTAgent - JSON process. Is it correct?

File: lib/fiware-iotagent-lib.js (hunk 7 above)
Comment (flopezag) [reply lib/fiware-iotagent-lib.js:256]: Agree, the reason was that at the beginning I did not covered all casuistic with the unit test so I prefer to make a clear comparison to true just to be sure that any other value except true or false is false. But with the last changes in the commonConfig I think that I covered all the cases and now the multiCore will be true or false only and then we could delete comparison.

it is clear, true, redundant also true, style followed well apparently also true ;)

File: doc/installationguide.md (hunk 9 above)
Comment (flopezag) [reply doc/installationguide.md:169]: Yes, I prefer sorting for the purpose of a faster access to then. When I check it was difficult to find some variables due to it was not sorted.

File: test/unit/general/config-multi-core-test.js (hunk 10 above)
Comment (flopezag) [reply test/unit/general/config-multi-core-test.js]: Interesting...

You have discovered an issue almost on all the headers of the project...

File: test/unit/general/config-multi-core-test.js (hunk 8 above)
Comment (flopezag) [reply test/unit/general/config-multi-core-test.js:202]: mmm not completely sure. Making tests I had an issue with the numeric values. That what the reason to create this test. 

I do not care about others. I want that my one is completely covered. Checking the unit tests there were a few tests about env. variables.

File: lib/fiware-iotagent-lib.js (hunk 5 above)
Comment (jason-fox) [reply lib/fiware-iotagent-lib.js]: Yes - signature of 

```javascript
executeIoTAgent(newConfig, iotAgent , callback) {
```

Would allow the logging to be placed in each of the IoT Agents individually

**Review by jason-fox — COMMENTED** [review]
I'm basically happy, but the logging stuff should be left as the callback - this reduces the impact of the   change in the Agents themselves.

**Review by jason-fox — APPROVED** [review]
